File: exp6/xogame.py
import random

def print_board(board):
    for row in board:
        print(" | ".join(row))
        print("-" * 5)

def check_winner(board, player):
    for i in range(3):
        if all(board[i][j] == player for j in range(3)) or all(board[j][i] == player for j in range(3)):
            return True
    if all(board[i][i] == player for i in range(3)) or all(board[i][2-i] == player for i in range(3)):
        return True
    return False

def is_board_full(board):
    return all(board[i][j] != '-' for i in range(3) for j in range(3))

def get_empty_cells(board):
    return [(i, j) for i in range(3) for j in range(3) if board[i][j] == '-']

def user_move(board):
    while True:
        try:
            row = int(input("Enter row (1-3): ")) - 1
            col = int(input("Enter column (1-3): ")) - 1
            if board[row][col] == '-':
                return row, col
            else:
                print("That cell is already occupied. Try again.")
        except (ValueError, IndexError):
            print("Invalid input. Please enter a number between 1 and 3.")

def computer_move(board):
    empty_cells = get_empty_cells(board)
    best_move = None
    best_score = -float('inf')

    for row, col in empty_cells:
        board[row][col] = 'O'
        score = minimax(board, 0, False)
        board[row][col] = '-'
        if score > best_score:
            best_score = score
            best_move = (row, col)

    return best_move

def minimax(board, depth, isMaximizingPlayer):
    result = evaluate(board)
    if result is not None:
        return result

    if isMaximizingPlayer:
        bestVal = -float('inf')
        for i in range(3):
            for j in range(3):
                if board[i][j] == '-':
                    board[i][j] = 'O'
                    value = minimax(board, depth + 1, False)
                    board[i][j] = '-'
                    bestVal = max(bestVal, value)
        return bestVal
    else:
        bestVal = float('inf')
        for i in range(3):
            for j in range(3):
                if board[i][j] == '-':
                    board[i][j] = 'X'
                    value = minimax(board, depth + 1, True)
                    board[i][j] = '-'
                    bestVal = min(bestVal, value)
        return bestVal

def evaluate(board):
    if check_winner(board, 'X'):
        return -1
    elif check_winner(board, 'O'):
        return 1
    elif is_board_full(board):
        return 0
    else:
        return None

def play_game():
    board = [['-' for _ in range(3)] for _ in range(3)]
    print("Welcome to Tic-Tac-Toe!")
    print_board(board)

    while True:
        # User's turn
        print("\nYour turn (X)")
        row, col = user_move(board)
        board[row][col] = 'X'
        print_board(board)
        if check_winner(board, 'X'):
            print("Congratulations! You win!")
            break
        if is_board_full(board):
            print("It's a draw!")
            break

        # Computer's turn
        print("\nComputer's turn (O)")
        row, col = computer_move(board)
        board[row][col] = 'O'
        print_board(board)
        if check_winner(board, 'O'):
            print("Sorry, you lose!")
            break
        if is_board_full(board):
            print("It's a draw!")
            break

    print("Thanks for playing!")

play_game()

File: exp6/xogame.c
#include <stdio.h>

char board[3][3];

void print_board() {
    for (int i = 0; i < 3; i++) {
        printf(" %c | %c | %c\n", board[i][0], board[i][1], board[i][2]);
        if (i < 2) printf("-----------\n");
    }
}

int evaluate(char player) {
    for (int i = 0; i < 3; i++)
        if (board[i][0] == player && board[i][1] == player && board[i][2] == player)
            return 1;

    for (int i = 0; i < 3; i++)
        if (board[0][i] == player && board[1][i] == player && board[2][i] == player)
            return 1;

    if ((board[0][0] == player && board[1][1] == player && board[2][2] == player) ||
        (board[0][2] == player && board[1][1] == player && board[2][0] == player))
        return 1;

    return 0;
}

int is_board_full() {
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            if (board[i][j] == ' ')
                return 0;

    return 1;
}

int minimax(char player) {
    if (evaluate('X')) return 1;
    if (evaluate('O')) return -1;
    if (is_board_full()) return 0;

    int best = (player == 'X') ? -1000 : 1000;

    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            if (board[i][j] == ' ') {
                board[i][j] = player;
                int score = minimax((player == 'X') ? 'O' : 'X');
                board[i][j] = ' ';
                if ((player == 'X' && score > best) || (player == 'O' && score < best))
                    best = score;
            }

    return best;
}

void make_ai_move() {
    int best = -1000, row, col;

    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            if (board[i][j] == ' ') {
                board[i][j] = 'X';
                int score = minimax('O');
                board[i][j] = ' ';
                if (score > best) {
                    best = score;
                    row = i;
                    col = j;
                }
            }

    board[row][col] = 'X';
}

int main() {
    printf("Welcome to Tic-Tac-Toe!\n");
    printf("You are playing as O, and the computer is playing as X.\n");
    printf("Enter your moves using row and column numbers (e.g., 1 1 for the top-left corner).\n");

    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            board[i][j] = ' ';

    while (1) {
        print_board();
        int row, col;
        scanf("%d %d", &row, &col);
        row--; col--;
       
        if (board[row][col] != ' ') {
            printf("That cell is already occupied! Try again.\n");
            continue;
        }
       
        board[row][col] = 'O';
       
        if (evaluate('O')) {
            print_board();
            printf("Congratulations! You win!\n");
            break;
        }
       
        if (is_board_full()) {
            print_board();
            printf("It's a tie!\n");
            break;
        }
       
        make_ai_move();
       
        if (evaluate('X')) {
            print_board();
            printf("The computer wins!\n");
            break;
        }
       
        if (is_board_full()) {
            print_board();
            printf("It's a tie!\n");
            break;
        }
    }

    return 0;
}
